text-feature-translation/src/main/java/net/kyori/adventure/text/feature/translation/TranslationRegistryImpl.java

```diff
@@ -0,0 +1,125 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.feature.translation;
+
+import net.kyori.adventure.key.DuplicateKeyException;
+import net.kyori.adventure.key.Key;
+import net.kyori.examination.Examinable;
+import net.kyori.examination.ExaminableProperty;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.text.MessageFormat;
+import java.util.Comparator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.stream.Stream;
+
+/* package */ final class TranslationRegistryImpl extends TranslatableComponentRenderer implements TranslationRegistry {
+  /* package */ static final TranslationRegistry INSTANCE = new TranslationRegistryImpl();
+
+  private static final class Translation implements Key, Examinable {
+    private final String owner;
+    private final String key;
+    private final Map<Locale, MessageFormat> formats;
+
+    private Translation(final @NonNull Key key) {
+      this.owner = key.namespace();
+      this.key = key.value();
+      this.formats = new ConcurrentSkipListMap<>(Comparator.comparing(Locale::toLanguageTag));
+    }
+
+    private void register(final @NonNull Locale locale, final @NonNull MessageFormat format) {
+      this.formats.put(locale, format);
+    }
+
+    private MessageFormat translate(final @NonNull Locale locale) {
+      return this.formats.get(locale);
+    }
+
+    @Override
+    public @NonNull String namespace() {
+      return this.owner;
+    }
+
+    @Override
+    public @NonNull String value() {
+      return this.key;
+    }
+
+    @Override
+    public @NonNull String asString() {
+      return this.value();
+    }
+
+    @Override
+    public int compareTo(final @NonNull Key key) {
+      return this.value().compareTo(key.value());
+    }
+
+    @Override
+    public @NonNull Stream<? extends ExaminableProperty> examinableProperties() {
+      return Stream.of(
+        ExaminableProperty.of("owner", this.owner),
+        ExaminableProperty.of("key", this.key),
+        ExaminableProperty.of("formats", this.formats)
+      );
+    }
+  }
+
+  private final Map<String, Translation> translations = new ConcurrentSkipListMap<>();
```

Comment:
Does this have to be a `ConcurrentSkipListMap`? This could just as easily been a `ConcurrentHashMap` (I assume you meant to make this class thread-safe).